src/cargo/util/auth.rs

```diff
@@ -0,0 +1,427 @@
+//! Registry authentication support.
+
+use crate::util::{config, CargoResult, Config};
+use anyhow::{bail, format_err, Context as _};
+use cargo_util::ProcessError;
+use core::fmt;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::error::Error;
+use std::io::{Read, Write};
+use std::path::PathBuf;
+use std::process::{Command, Stdio};
+use url::Url;
+
+use crate::core::SourceId;
+use crate::ops::RegistryCredentialConfig;
+
+/// Get the credential configuration for a `SourceId`.
+pub fn registry_credential_config(
+    config: &Config,
+    sid: &SourceId,
+) -> CargoResult<RegistryCredentialConfig> {
+    #[derive(Deserialize)]
+    #[serde(rename_all = "kebab-case")]
+    struct RegistryConfig {
+        index: Option<String>,
+        token: Option<String>,
+        credential_process: Option<config::PathAndArgs>,
+    }
+
+    log::trace!("loading credential config for {}", sid);
+    if !sid.is_remote_registry() {
+        bail!(
+            "{} does not support API commands.\n\
+             Check for a source-replacement in .cargo/config.",
+            sid
+        );
+    }
+
+    // Handle crates.io specially, since it uses different configuration keys.
+    if sid.is_default_registry() {
+        config.check_registry_index_not_set()?;
+        let process = config.get::<Option<config::PathAndArgs>>("registry.credential-process")?;
+        let credential = if process.is_some() && config.cli_unstable().credential_process {
+            let process = process.unwrap();
+            RegistryCredentialConfig::Process((process.path.resolve_program(config), process.args))
+        } else if let Some(token) = config.get_string("registry.token")?.map(|p| p.val) {
```

Comment:
The old code has a blocker for the case when crates.io has both a credential process and a token. https://github.com/rust-lang/cargo/blob/0f75da631b4284180232c5a50bdf37e714acd059/src/cargo/ops/registry.rs#L404 I am not seeing the corresponding check here.
If this has slipped through, we probably need a test for this case.

Reply:
Hmm. `registry.credential-process`, is used as both as the default credential process for alternative registries, and the credential process for crates.io.

I wonder if we should separate these? For now, I think we just need to add the error (and a test) back as you're suggesting here. But it's something to think about.

Reply:
In the process of re-basing my asymmetric token work on yours I isolated my fix for this problem,
https://github.com/rust-lang/cargo/commit/1406b798d9b680b2e0d1dc3df6472c8a6f208589
If that code is useful use it. If not ignore it.

Reply:
Thanks! I've taken your fix to the issue.